cpp/dolfinx/io/checkpointing.cpp

```diff
@@ -0,0 +1,354 @@
+    engine->Put(var_num_dofs_per_cell, num_dofs_per_cell);
+    engine->Put(var_input_global_indices, input_global_indices_span.data());
+    engine->Put(var_x, mesh_x.subspan(0, num_vertices_local * 3).data());
+    engine->Put(var_topology_array, array_global.data());
+    engine->Put(var_topology_offsets, offsets_global.data());
+    engine->EndStep();
+  }
+}
+
+//-----------------------------------------------------------------------------
+/// @cond
+template void write_mesh<float>(ADIOS2Wrapper& ADIOS2,
+                                dolfinx::mesh::Mesh<float>& mesh);
+
+template void write_mesh<double>(ADIOS2Wrapper& ADIOS2,
+                                 dolfinx::mesh::Mesh<double>& mesh);
+
+/// @endcond
+
+//-----------------------------------------------------------------------------
+template <std::floating_point T>
+dolfinx::mesh::Mesh<T> read_mesh(ADIOS2Wrapper& ADIOS2, MPI_Comm comm)
+{
+
+  int rank, size;
+  MPI_Comm_rank(comm, &rank);
+  MPI_Comm_size(comm, &size);
+
+  auto io = ADIOS2.io();
+  auto engine = ADIOS2.engine();
+
+  engine->BeginStep();
+
+  // Attributes
+  std::string name;
+  std::int32_t dim, degree;
+  mesh::CellType cell_type;
+  basix::element::lagrange_variant lagrange_variant;
+
+  // Read attributes
+  {
+    adios2::Attribute<std::string> var_name
+        = io->InquireAttribute<std::string>("name");
+    adios2::Attribute<std::int32_t> var_dim
+        = io->InquireAttribute<std::int32_t>("dim");
+    adios2::Attribute<std::string> var_cell_type
+        = io->InquireAttribute<std::string>("cell_type");
+    adios2::Attribute<std::int32_t> var_degree
+        = io->InquireAttribute<std::int32_t>("degree");
+    adios2::Attribute<std::string> var_variant
+        = io->InquireAttribute<std::string>("lagrange_variant");
+
+    name = var_name.Data()[0];
+    dim = var_dim.Data()[0];
+    cell_type = mesh::to_type(var_cell_type.Data()[0]);
+    degree = var_degree.Data()[0];
+    lagrange_variant = string_to_variant[var_variant.Data()[0]];
+  }
+
+  // Scalar variables
+  std::uint64_t num_vertices_global;
+  std::uint64_t num_cells_global;
+  std::uint32_t num_dofs_per_cell;
+
+  // Read scalar variables
+  {
+    adios2::Variable<std::uint64_t> var_num_vertices
+        = io->InquireVariable<std::uint64_t>("num_vertices");
+    adios2::Variable<std::uint64_t> var_num_cells
+        = io->InquireVariable<std::uint64_t>("num_cells");
+    adios2::Variable<std::uint32_t> var_num_dofs_per_cell
+        = io->InquireVariable<std::uint32_t>("num_dofs_per_cell");
+
+    engine->Get(var_num_vertices, num_vertices_global);
+    engine->Get(var_num_cells, num_cells_global);
+    engine->Get(var_num_dofs_per_cell, num_dofs_per_cell);
+  }
+
+  // Compute local sizes, offsets
+  std::array<std::int64_t, 2> _local_range
+      = dolfinx::MPI::local_range(rank, num_vertices_global, size);
+
+  std::array<std::uint64_t, 2> local_range{(std::uint64_t)_local_range[0],
```

Comment:
What is happening here if the size of the MPI communicator has changed?

Reply:
Determining the sizes for each processor in a balanced manner. 
The existing function is based on `int64_t` and a type cast is don't here. 

The existing function `local_range` can be templated.

Reply:
A new function is added similar to local_range to compute offset and local size directly.